From d1bc3fa3c704bc233f0036ea13204d7790d26477 Mon Sep 17 00:00:00 2001
From: Jan de Muijnck-Hughes <jfdm@st-andrews.ac.uk>
Date: Mon, 5 Dec 2016 11:17:51 +0000
Subject: [PATCH] Make fetching of data files more robust. Fixes #3544

`getDataFileName` didn't take into account possible changes in Idris' datadir that were not cabal made. This fix adds a more robust variant of `getDataFileName`.
---
 src/IRTS/System.hs    | 7 ++++++-
 src/Idris/IdrisDoc.hs | 4 ++--
 src/Idris/Info.hs     | 5 +++++
 3 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/src/IRTS/System.hs b/src/IRTS/System.hs
index def563e..9d3d11a 100644
--- a/src/IRTS/System.hs
+++ b/src/IRTS/System.hs
@@ -6,7 +6,7 @@ License     : BSD3
 Maintainer  : The Idris Community.
 -}
 {-# LANGUAGE CPP #-}
-module IRTS.System( getDataFileName
+module IRTS.System( getIdrisDataFileByName
                   , getCC
                   , getLibFlags
                   , getIdrisDataDir
@@ -42,6 +42,11 @@ getIdrisDataDir = do
       return ddir
     Just ddir -> return ddir
 
+getIdrisDataFileByName :: String -> IO FilePath
+getIdrisDataFileByName fn = do
+  dir <- getIdrisDataDir
+  return $ dir </> fn
+
 overrideIdrisSubDirWith :: String  -- ^ Sub directory in `getDataDir` location.
                         -> String  -- ^ Environment variable to get new location from.
                         -> IO FilePath
diff --git a/src/Idris/IdrisDoc.hs b/src/Idris/IdrisDoc.hs
index 012554f..6c144ae 100644
--- a/src/Idris/IdrisDoc.hs
+++ b/src/Idris/IdrisDoc.hs
@@ -19,7 +19,7 @@ import Idris.Docs
 import Idris.Docstrings (nullDocstring)
 import qualified Idris.Docstrings as Docstrings
 import Idris.Parser.Helpers (opChars)
-import IRTS.System (getDataFileName)
+import IRTS.System (getIdrisDataFileByName)
 
 import Control.Applicative ((<|>))
 import Control.Monad (forM_)
@@ -708,5 +708,5 @@ copyDependencies :: FilePath -- ^ The base directory to which
                              --   dependencies should be written
                  -> IO ()
 copyDependencies dir =
-  do styles <- getDataFileName $ "idrisdoc" </> "styles.css"
+  do styles <- getIdrisDataFileByName $ "idrisdoc" </> "styles.css"
      copyFile styles (dir </> "styles.css")
diff --git a/src/Idris/Info.hs b/src/Idris/Info.hs
index 4ffe3d4..eb77a8c 100644
--- a/src/Idris/Info.hs
+++ b/src/Idris/Info.hs
@@ -22,6 +22,7 @@ module Idris.Info
   , getIdrisHistoryFile
   , getIdrisInstalledPackages
   , getIdrisLoggingCategories
+  , getIdrisDataFileByName
   ) where
 
 import Idris.AbsSyntax (loggingCatsStr)
@@ -34,6 +35,7 @@ import Version_idris (gitHash)
 import Data.Version
 import System.Directory
 import System.FilePath
+
 getIdrisDataDir :: IO String
 getIdrisDataDir = S.getIdrisDataDir
 
@@ -88,3 +90,6 @@ getIdrisInstalledPackages = installedPackages
 
 getIdrisLoggingCategories :: IO [String]
 getIdrisLoggingCategories = return $ words loggingCatsStr
+
+getIdrisDataFileByName :: String -> IO FilePath
+getIdrisDataFileByName = S.getIdrisDataFileByName
